src/serializer/ResidualHeapSerializer.js

```diff
@@ -731,7 +748,14 @@ export class ResidualHeapSerializer {
       if (s === "GLOBAL") return undefined;
       invariant(s instanceof FunctionValue);
       if (this.additionalFunctionGenerators.has(s)) {
-        if (additionalFunction !== undefined && additionalFunction !== s) return undefined;
+        if (additionalFunction !== undefined) {
+          if (this.isNestedOptimizedFunction(s, additionalFunction)) {
```

Comment:
If we first visit the scope of the parent and then alter a nested function, then this will happy continue (and eventually return the parent).

If we first visit the scope of the nested function, and then the parent, then this will return undefined.

This non-deterministic behavior is bad.

This function used to return the only referencing function, if any. What do you want to happen/allow in the case of nesting?

Reply:
See the test cases, I need those to work :) sorry I’m on mobile atm but can explain better when I’m back to my desk.

Reply:
Do I understand your comment correctly that `scopes` are not guaranteed to be in any specific order?
Or is the problem in something else?

Reply:
Right, the scopes would be listed whatever order the visitor discovered them (sets retain insertion order), but there's no guaranteed order here.

Reply:
@gaearon
You misunderstood my reply to Nikolai. I wasn’t suggesting that the tests were good indicators of the problem. I was saying that they demonstrate the problem in case he wanted to know what this PR was attempting to fix

Reply:
Got it, thanks for explaining!

Reply:
I think I addressed your issue Nikolai, I've changed the logic. Just adding a test for it too.